Check if properRoute is found before acting on it

kdhira opened the pull request:
In the cases where the request route is not matched to any Koa route,
bail out before trying to format the path string.

Issue: #59

kdhira committed:
Check if properRoute is found before acting on it

In the cases where the request route is not matched to any Koa route,
bail out before trying to format the path string.

Issue: #59

kdhira commented on src/koa-middleware.js:
```diff
@@ -93,6 +93,12 @@ class KoaMiddleware {
+        // If proper route is not found, send an undefined route
+        // The caller is responsible for setting a default "N/A" route in this case
+        if (!properRoute) {
+            return undefined;
```
In the `_handleResponse` method, this will set the route label to 'N/A' though the `||` operator. Let me know if there is a cleaner solution

kdhira commented:
@kobik can you take a look at this?

kobik commented:
Hi @kdhira , thanks for the PR!

I'll have a look tomorrow.

kdhira commented:
Thanks @kobik

Let me know if there's any questions or anything you're unsure about. Like I mentioned before, the fix minimally circumvents the NPE which gets the route label working with just 'N/A' for my usecase.

Also I'm hoping that these changes can be released to NPM as a version of prometheus-api-metrics, probably something like 3.2.1 of the package. Is there anything extra I need to do (ie version bumping etc) or is that all handled by you?

kobik reviewed (changes requested):
Thanks @kdhira.
Is it possible to add a test for this scenario?

kdhira committed:
Add test for koa middleware unmatched route scenario

Issue: #59

kdhira commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "prometheus-api-metrics",
-  "version": "0.0.0",
+  "version": "3.2.0",
```
I found this to be the problem with package-lock.json file, giving me a version difference. I restored the `3.2.0` version to the package.json

Also I had to use Node v15 with the v2 lockfileversion otherwise it would try to restore the v1 config, rewriting most of the package-lock.json

kobik replied:
the version should remain 0.0.0 as we use `semantic-release` to generate the next version which sets it in the package.json file as part of the release process

kdhira replied:
Ah okay, thanks for explaining this. This is easy for me to revert, will switch it back

kdhira replied:
I have reverted the package version to `0.0.0`

kdhira commented on test/unit-test/metric-middleware-koa-test.js:
```diff
@@ -653,6 +653,57 @@ describe('metrics-middleware', () => {
                 Prometheus.register.clear();
             });
         });
+        describe('when using middleware request and path not matched to any specific route', () => {
```
I had to spend a little time to understand how the tests were made and what level of things were being tested. So I believe this 1 unit test is sufficient for the code changes I've made. But please advise if there is anything else you request

kdhira commented:
Hi @kobik,

I added a unit test for this scenario. Please let me know if anything is wrong. I have a commit for updating to dependencies from `npm audit fix`.

kobik reviewed (changes requested):
@kdhira thanks for your PR.

See my comment

kdhira commented:
Thanks @kobik,

I've seen your comment regarding the package version, and have reverted it. I had this as a separate commit so it was simple for me to just rebase it out.

Let me know if there's anything else I should change :)

kobik commented:
@kdhira, can you also revert the package-lock.json changes?

I'd like to handle them in a separate PR.

kdhira commented:
@kobik, I have dropped my commit that updated dependencies (and doing so reverted the package-lock.json changes I made). Hope this helps!

kobik commented:
Sure, thanks again @kdhira !

kobik commented:
@kdhira it seems like the new test didn't cover our case.

see https://coveralls.io/builds/47157280/source?filename=src%2Fkoa-middleware.js#L99

kdhira commented:
Hmm I'll have to look at the test again, but I'm pretty confident that line gets run as the test I made has to resolve the route to use in the metric as 'N/A' somehow (through the `|| 'N/A'` in `_handleResponse`) ...

kdhira commented:
Does the coveralls tool run the unit-tests or the integration tests (or both) to detect coverage?

kdhira committed:
Add integration test for wildcard path scenario

Issue: #59

kdhira commented:
Hey @kobik,

I think I have done what I need to ensure coverage. I added an integration test, and updated the test koa-server implementation so that the wildcard endpoint case could be correctly tested.

I also ran `npm run integration-tests` and after this addition it no longer shows line 99 as being an Uncovered Line.

Should be ready to run the Github workflows again

kobik commented:
you're correct, for some reason it's only collected from integration tests.
you can update the workflow to run `nyc npm test` instead of running unit and integration separately, which should collect the coverage of both of them.

but also what you did should be fine 👍
